agent/reconnectingpty/buffered.go

```diff
@@ -0,0 +1,167 @@
+	cmd.Env = append(b.command.Env, "TERM=xterm-256color")
+	cmd.Dir = b.command.Dir
+	ptty, process, err := pty.Start(cmd)
+	if err != nil {
+		return err
+	}
+	b.ptty = ptty
+	b.process = process
+
+	// Multiplex the output onto the circular buffer and each active connection.
+	//
+	// We do not need to separately monitor for the process exiting.  When it
+	// exits, our ptty.OutputReader() will return EOF after reading all process
+	// output.
+	go func() {
+		buffer := make([]byte, 1024)
+		for {
+			read, err := ptty.OutputReader().Read(buffer)
+			if err != nil {
+				// When the PTY is closed, this is triggered.
+				// Error is typically a benign EOF, so only log for debugging.
+				if errors.Is(err, io.EOF) {
+					logger.Debug(ctx, "unable to read pty output, command might have exited", slog.Error(err))
+				} else {
+					logger.Warn(ctx, "unable to read pty output, command might have exited", slog.Error(err))
+					b.metrics.WithLabelValues("output_reader").Add(1)
+				}
+				// Could have been killed externally or failed to start at all (command
+				// not found for example).
+				b.closeSession("unable to read pty output, command might have exited")
+				break
+			}
+			part := buffer[:read]
+			b.mutex.Lock()
+			_, err = b.circularBuffer.Write(part)
+			if err != nil {
+				logger.Error(ctx, "write to circular buffer", slog.Error(err))
+				b.metrics.WithLabelValues("write_buffer").Add(1)
+			}
+			for cid, conn := range b.activeConns {
+				_, err = conn.Write(part)
+				if err != nil {
+					logger.Warn(ctx,
+						"error writing to active connection",
+						slog.F("other_conn_id", cid),
+						slog.Error(err),
+					)
+					b.metrics.WithLabelValues("write").Add(1)
+				}
+			}
+			b.mutex.Unlock()
+		}
+	}()
+
+	return nil
+}
+
+// attach attaches to the pty and replays the buffer.  If the context closes it
+// will detach the connection but leave the process up.  A connection ID is
+// required so that logs in the pty goroutine can reference the same ID
+// reference in logs output by each individual connection when acting on those
+// connections.
+func (b *bufferedBackend) attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) (pty.PTYCmd, error) {
+	// Resize the PTY to initial height + width.
+	err := b.ptty.Resize(height, width)
+	if err != nil {
+		// We can continue after this, it's not fatal!
+		logger.Warn(ctx, "reconnecting PTY initial resize failed, but will continue", slog.Error(err))
+		b.metrics.WithLabelValues("resize").Add(1)
+	}
+
+	// Write any previously stored data for the TTY and store the connection for
+	// future writes.
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	prevBuf := slices.Clone(b.circularBuffer.Bytes())
+	_, err = conn.Write(prevBuf)
+	if err != nil {
+		b.metrics.WithLabelValues("write").Add(1)
+		return nil, xerrors.Errorf("write buffer to conn: %w", err)
+	}
+	b.activeConns[connID] = conn
```

Comment:
there is a race condition here:  the command may have already exited and run its routine that closes all connections.  Then, we come in and add a new connection, which will just hang forever if there is no input.

We need to store lifecycle information that we can check while holding a mutex to decide whether the pty is still open.

Reply:
I think I could also fix this if instead of using a map I used channels (send pty output to a channel, select on that in `attach` and write to the conn).

But even if that works, do you think tracking the state in the upper layer is still weird enough that I should still move it to the backend?  Lifecycle is almost entirely what the upper layer does so I am wondering if I just get rid of that layer entirely.

Hmmmm yeah I think I will do that, the backend really should understand its own state even if this particular bug goes away, but let me know if this seems wrong.  Will probably just keep `New` as-is and move the rest to the individual backends.

Reply:
Refactored!  I think I was prematurely abstracting before.  There is some duplication we might want to examine but I think overall things make more sense now.